File: GeeksforGeeks/C++/Medium/Water the plants/water-the-plants.cpp
//{ Driver Code Starts
// driver code

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution{
    public:
    int min_sprinklers(int gallery[], int n)
    {
        // code here
        vector<pair<int, int>> v;
        for (int i = 0; i < n; i++) {
            if (gallery[i] == -1)
                continue;
            v.push_back({
                max(0, i - gallery[i]),
                min(n - 1, i + gallery[i])
            });
        }
        sort(v.begin(), v.end());
        int start = 0, i = 0, ans = 0;
        while (start < n) {
            int mx = INT_MIN;
            while (i < v.size()) {
                if (v[i].first > start)
                    break;
                mx = max(mx, v[i].second);
                i++;
            }
            if (mx < start)
                return -1;
            ans++;
            start = mx + 1;
        }
        return ans;
    }
};


//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        int n;
        cin>>n;
        
        int gallery[n];
        for(int i=0; i<n; i++)
            cin>> gallery[i];
        Solution obj;
        cout<< obj.min_sprinklers(gallery,n) << endl;
    }
	return 1;
}

// } Driver Code Ends

File: GeeksforGeeks/Python3/Medium/Water the plants/water-the-plants.py
#User function Template for python3
class Solution:
    def min_sprinklers(self, gallery, n):
        # code here
        water = []
        for i in range(n):
            if gallery[i] != -1:
                water.append((i-gallery[i], i+gallery[i]))
        if len(water) == 0:
            return -1
        water.sort()
        ans = 0
        left, right = 0, 0
        if water[0][0] > right:
            return -1
        else:
            ans = 1
            right = water[0][1]
        for i in range(1, len(water)):
            if right >= n-1:
                break
            if water[i][0] <= left:
                right = max(right, water[i][1])
            elif water[i][0] > right+1:
                return -1
            else:
                ans += 1
                left = right+1
                right = water[i][1]
        return ans if right >= n-1 else -1
#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input())
        gallery = [int(x) for x in input().strip().split()]
        obj = Solution()
        print(obj.min_sprinklers(gallery,n))

# } Driver Code Ends